modules/xfem/src/base/XFEM.C

```diff
@@ -1507,3 +1557,162 @@ XFEM::getXFEMqRuleOnSurface(std::vector<Point> & intersection_points, std::vecto
     }
   }
 }
+
+void
+XFEM::saveSolutionForNode(const Node * node_to_store_to,
```

Comment:
There is a lot of code duplication in these next four methods. Have you considered breaking this into a few more methods and dropping all of the duplicate code?